Chaincase.Common/Services/AesThenHmac.cs

```diff
@@ -0,0 +1,402 @@
+                for (var i = 0; i < sentTag.Length; i++)
+                    compare |= sentTag[i] ^ calcTag[i];
+
+                //if message doesn't authenticate return null
+                if (compare != 0)
+                    return null;
+
+                using (var aes = new AesManaged
+                {
+                    KeySize = KeyBitSize,
+                    BlockSize = BlockBitSize,
+                    Mode = CipherMode.CBC,
+                    Padding = PaddingMode.PKCS7
+                })
+                {
+
+                    //Grab IV from message
+                    var iv = new byte[ivLength];
+                    Array.Copy(encryptedMessage, nonSecretPayloadLength, iv, 0, iv.Length);
+
+                    using (var decrypter = aes.CreateDecryptor(cryptKey, iv))
+                    using (var plainTextStream = new MemoryStream())
+                    {
+                        using (var decrypterStream = new CryptoStream(plainTextStream, decrypter, CryptoStreamMode.Write))
+                        using (var binaryWriter = new BinaryWriter(decrypterStream))
+                        {
+                            //Decrypt Cipher Text from Message
+                            binaryWriter.Write(
+                              encryptedMessage,
+                              nonSecretPayloadLength + iv.Length,
+                              encryptedMessage.Length - nonSecretPayloadLength - iv.Length - sentTag.Length
+                            );
+                        }
+                        //Return Plain Text
+                        return plainTextStream.ToArray();
+                    }
+                }
+            }
+        }
+
+        public static byte[] EncryptWithPassword(byte[] secretMessage, string password, byte[] nonSecretPayload = null)
+        {
+            nonSecretPayload = nonSecretPayload ?? new byte[] { };
+
+            //User Error Checks
+            if (secretMessage == null || secretMessage.Length == 0)
+                throw new ArgumentException("Secret Message Required!", "secretMessage");
+
+            var payload = new byte[((SaltBitSize / 8) * 2) + nonSecretPayload.Length];
+
+            Array.Copy(nonSecretPayload, payload, nonSecretPayload.Length);
+            int payloadIndex = nonSecretPayload.Length;
+
+            byte[] cryptKey;
+            byte[] authKey;
+            //Use Random Salt to prevent pre-generated weak password attacks.
+            using (var generator = new Rfc2898DeriveBytes(password, SaltBitSize / 8, Iterations))
+            {
+                var salt = generator.Salt;
+
+                //Generate Keys
+                cryptKey = generator.GetBytes(KeyBitSize / 8);
+
+                //Create Non Secret Payload
+                Array.Copy(salt, 0, payload, payloadIndex, salt.Length);
+                payloadIndex += salt.Length;
+            }
+
+            //Deriving separate key, might be less efficient than using HKDF, 
```

Comment:
KDFs should be as inefficient as possible :)

Reply:
Yeah, I'm not sure what they og author Tuley means by this exactly. Maybe that's the reasoning for deriving separate salts